javascript/stimulus_reflex.js

```diff
@@ -246,18 +249,46 @@ const setupDeclarativeReflexes = () => {
         if (!actions.includes(action)) actions.push(action)
       } else {
         action = `${reflex.split('->')[0]}->stimulus-reflex#__perform`
-        if (!controllers.includes('stimulus-reflex'))
+        if (!controllers.includes('stimulus-reflex')) {
           controllers.push('stimulus-reflex')
+        }
         if (!actions.includes(action)) actions.push(action)
       }
     })
-    if (controllers.length > 0)
+    if (controllers.length > 0) {
       element.setAttribute('data-controller', controllers.join(' ').trim())
-    if (actions.length > 0)
+    }
+    if (actions.length > 0) {
       element.setAttribute('data-action', actions.join(' ').trim())
+    }
   })
 }
 
+// compute the DOM element(s) which will be the morph root
+// use the data-reflex-root attribute on the reflex or the controller
+// optional value is a CSS selector(s); comma-separated list
+// order of preference is data-reflex, data-controller, document body (default)
+const getReflexRoots = (element, target) => {
+  let reflexRoot = element.dataset.reflexRoot
+  if (reflexRoot !== undefined && reflexRoot.length) {
+    return reflexRoot
+      .split(',')
+      .filter(f => document.querySelector)
+      .join(',')
+  } else {
+    const controllerElement = findReflexController(element, target).element
+    reflexRoot = controllerElement.dataset.reflexRoot
+    if (reflexRoot !== undefined && reflexRoot.length) {
+      return reflexRoot
+        .split(',')
+        .filter(f => document.querySelector)
+        .join(',')
+    } else {
+      return 'body'
+    }
+  }
+}
+
```

Comment:
This is great. Looks like we still have some JavaScript formatting discrepancies between our systems.

Reply:
I'll run it through the command line tool before I commit!

Reply:
```
pete:~/stimulus_reflex/javascript (selectors)* $ yarn run prettier-standard --check ./stimulus_reflex.js
yarn run v1.12.3
$ /home/pete/stimulus_reflex/javascript/node_modules/.bin/prettier-standard --check ./stimulus_reflex.js
All matched files use Standard code style!
Done in 0.49s.
```